torch/csrc/dynamo/framelocals_mapping.h

```diff
@@ -3,16 +3,64 @@
 #include <torch/csrc/utils/python_compat.h>
 
 #ifdef __cplusplus
+
+#include <string>
+#include <unordered_map>
+
+#include <torch/csrc/dynamo/utils.h>
+#include <torch/csrc/utils/pybind.h>
+
 extern "C" {
+
+typedef struct VISIBILITY_HIDDEN FrameLocalsMapping {
+ private:
+  std::unordered_map<std::string, PyObject*> _map;
```

Comment:
Is a `std::unordered_map` actually much faster than a Python dict?

The thing I was suggesting in #93753 was replacing:
```
f_locals["foo"]
```
with
```
fastlocals[3]
```
where the conversion to figure out that the variable `"foo"` is stored at index `3` can be done ahead of time.  Thus making the loop just a single load instruction (rather than a hashtable lookup).

Since both PyDict and unordered_map are hashtables, this seems worse than computing the fastlocals offset ahead of time (and storing it on the guard).

Reply:
nit: you could use pybind11 to do this casting for you?

Reply:
I don't believe there is a guarantee on the ordering of the fastlocals names (but in practice, this may be true). The safe thing to do would be to add a check that fastlocals name order is the same.

I'd also think that a C++ unordered has less overhead than using Python dict, although I'd have to measure to be sure. The point of this PR is not to introduce the complete optimization, but rather, is a step to moving framelocals/fastlocals lookup to C++. I plan on incrementally introducing optimizations in followup PRs. For this PR, we just need to make sure there are no regressions.

Reply:
> I don't believe there is a guarantee on the ordering of the fastlocals names (but in practice, this may be true). The safe thing to do would be to add a check that fastlocals name order is the same.

Isn't the order defined on the code object which is a constant?  How could the order change?

Reply:
Ah, I hadn't considered that.

Reply:
I think directly using fastlocals will actually be easier to implement than a unordered_map.  This also matches what CPython does, I believe:
`LOAD_FAST 3` is just `push(fastlocals[3])`